Amendment RPC enhancements

nbougalis opened the pull request:
* RPC command to veto/unveto
* Store votes
* Add vote information to JSON
* Add ledger majority information to JSON
* Config section for vetos

Reviews: @seelabs, @JoelKatz

seelabs commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -26,36 +26,162 @@
+            "([abcdefABCDEF0-9]*)" // <hex amend ID>
+            "(?:\\s*)"             // whitespace (optional)
+            "(?:.*\\S+)"           // <value>
+        , boost::regex_constants::optimize
```
I don't think this regular expression works:

1) The `value` group is never captured, and it is referred to in the code below.
2) If we remove the `?:` and capture the value group, it captures too much, including the `=` delimiter between the key and value
3) It will match any delimiter will work, not just an `=`
4) I prefer `\s*` to `(?:\s*)`
5) Since we call `SetHexExact` below, we should constrain the key to the expected number of characters
5) Raw strings can get rid of some of the backslashes

How about something like: `R"(^\s*([a-fA-F0-9]{64})\s*=\s*(\S+))"`

nbougalis replied:
Actually, the comment is wrong. I should remove that. The expression captures any number of hex characters, and optionally, any data present after whitespace. I would consider changing the `*` in `([abcdefABCDEF0-9]*)` to a `+` or even specifying the exact number of characters needed.

seelabs commented on src/ripple/app/main/Amendments.cpp:
```diff
@@ -0,0 +1,48 @@
+{
+    return
+    {
+    };
```
Strange formatting, it looks like a block. I'd prefer `return {};`

nbougalis replied:
I can change it to be a bit more conventional, but the idea is that in the future we can have something like the following:

```return
{
    { "hexid name" },
    { "hexid name" }
};```

seelabs replied:
Yeah, I'd change it to be more conventional. When we add elements to the collection in the future we can change the formatting. Right now it's drawing undue attention to itself because of the formatting.

seelabs commented on src/ripple/app/misc/AmendmentTable.h:
```diff
@@ -221,21 +67,21 @@ class AmendmentTable
 
     virtual void
-    doValidatedLedger (LedgerIndex ledgerSeq, enabledAmendments_t enabled) = 0;
+    doValidatedLedger (LedgerIndex ledgerSeq, std::set <uint256> enabled) = 0;
```
Pass `enabled` by `const &`

nbougalis replied:
Done.

seelabs commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -26,54 +26,155 @@
+        if (!boost::regex_match (line, match, re1))
+        {
+            throw std::runtime_error (
+                "Invalid entry in [" + section.name () + "]");
```
Write out the line that we are trying to match so it's easy to find and correct in the config

seelabs commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -108,220 +207,168 @@ class AmendmentTableImpl final : public AmendmentTable
+        auto s = add (a.first);
+
+        if (!a.second.empty ())
+            s->name = a.second;
```
If we changed the param to `auto& a` we could `std::move (a.second)` here and save an allocation. OTOH, I'm less concerned about optimizations for on-off startup tasks like this. Your call.

seelabs commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -108,220 +207,168 @@ class AmendmentTableImpl final : public AmendmentTable
+        if (s)
+        {
+            if (!a.second.empty ())
+                s->name = a.second;
```
Could `std::move` here, see comment above.

seelabs commented on src/ripple/app/tests/AmendmentTable.test.cpp:
```diff
@@ -32,349 +33,299 @@ namespace ripple
-        TablePopulationAlgo populationAlgo = TablePopulationAlgo::addKnown)
+    static
+    Section
+    makeSection (std::vector<std::string> amendments)
```
pass by `const &`

nbougalis replied:
Done.

seelabs commented on src/ripple/app/tests/AmendmentTable.test.cpp:
```diff
@@ -32,349 +33,299 @@ namespace ripple
-        return amendmentNames;
+    static
+    Section
+    makeSection (uint256 amendment)
```
pass by `const &`

nbougalis replied:
Done.

seelabs commented on src/ripple/app/tests/AmendmentTable.test.cpp:
```diff
@@ -32,349 +33,299 @@ namespace ripple
-            for (auto const& i : badNumTokensConfigLines)
+        { // Two arguments are required - we pass three
+            Section test = section;
+            test.append (id + "Test Name");
```
I think this passes two arguments, not three (it probably passes the test because `Test` has non hex chars). We need to pass `" Test Name"` (notice the space before `Test`)

seelabs commented on src/ripple/app/tests/AmendmentTable.test.cpp:
```diff
@@ -678,25 +636,22 @@ class AmendmentTable_test final : public beast::unit_test::suite
                 validators, votes, ourVotes, enabled, majority);
 
-            if (i < 14)
+            if (i < 13)
```
Note: These numbers changed because we changed the way we calculate the threshold in `AmendmentTableImpl::doVoting` from rounding to floor.

seelabs commented on src/ripple/app/misc/AmendmentTableImpl.cpp:
```diff
@@ -418,36 +436,34 @@ AmendmentTableImpl::doVoting (
-        m_journal.trace <<
-            amendmentSet.mTrustedValidations << " trusted validations, threshold is "
-            << threshold;
+    vote->mThreshold = (vote->mTrustedValidations * mMajorityFraction) / 256;
```
Do we have to worry about `vote->mThreshold == 0`? I.e. an amendment passing with zero votes?

nbougalis replied:
How about:

    vote->mThreshold = std::max(1,
        vote->mTrustedValidations * mMajorityFraction) / 256);

seelabs commented:
:+1: I left two nit comments in the latest commits, but thumbs up with or without them.

JoelKatz commented:
:+1:

nbougalis commented:
I rebased, squashed everything together and added a bit of extra logging. I am removing the "Passed" label because I made a small tweak in the code. Previously, a single server could vote multiple times for the same amendment within a single round - obviously that's *not* what we want. This is no longer the case.

vinniefalco commented:
This decreases coverage :-1:

nbougalis commented:
Merged as f13668371e9b638cb2ffc274340048e66e6d5447
